### PR TITLE
Add qp option and test gate

### PR DESCRIPTION
**Pull request by davidhadas**
# Changes

- Added QPOption - a package to connect a security gate with Queue Proxy Option
- Added Tesg-gate - a test package 
- Added a diagram for Readme

The repository now includes:

- A TestGate (a test plug) and all the glue connecting it to be a Queue Proxy Option

### CODE REVIEW

**Comment by davidhadas** [comment]
@upodroid seems like the easycla issue is still on, any thoughts how to get this resolved?

@psschwei, @rhuss @maximilien  - this is the first batch of code and Readme. 
It includes:
- A TestGate (a test plug) and all the glue connecting it to be a Queue Proxy Option
- A full readme describing Guard (but no guard specific code). 


Now, before we dive into Guard's actual code, is a good point to review the structure of a queue proxy option (extension) and understand how a security gate is added to a Queue Proxy and how it is configured (maybe also play around with the test-gate and consider how it can be used for e2e testing). 

btw, unit tests seem to be working, maybe good to add coverage  - [Codecov](https://codecov.io/gh/knative/serving/pull/13146?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=knative) Report (not sure how we add it)

File: go.mod
Comment (upodroid) [inline go.mod:1]:
```suggestion
module knative.dev/security-guard
```

File: pkg/qpoption/qpoption.go
Comment (psschwei) [inline pkg/qpoption/qpoption.go:40]:
```suggestion
var qpextentionPrefix = "qpextention.knative.dev/"
```

File: pkg/qpoption/qpoption.go
Comment (psschwei) [inline pkg/qpoption/qpoption.go:75]: could we use the more idiomatic `if err = function; err != nil {}` form here?

File: pkg/qpoption/qpoption.go
Comment (psschwei) [inline pkg/qpoption/qpoption.go:141]:
```suggestion
	serviceName := defaults.Env.ServingService
```

File: pkg/qpoption/qpoption_test.go
Comment (psschwei) [inline pkg/qpoption/qpoption_test.go:323]: should we remove the comment?

File: pkg/test-gate/test-gate_test.go
Comment (psschwei) [inline pkg/test-gate/test-gate_test.go:31]:
```suggestion
type dLog struct {}
```

File: pkg/pluginterfaces/pluginterfaces_test.go
Comment (psschwei) [inline pkg/pluginterfaces/pluginterfaces_test.go:26]:
```suggestion
type plug struct {}
```

File: pkg/pluginterfaces/pluginterfaces.go
Comment (psschwei) [inline pkg/pluginterfaces/pluginterfaces.go:44]: should we use a more descriptive variable name for `c`?

File: pkg/qpoption/qpoption_test.go
Comment (psschwei) [inline pkg/qpoption/qpoption_test.go:187]:
```suggestion
	}{{
		name: "empty", 
		wantResp: true, 
		config: map[string]string{},
	},{
        	name: "few", 
        	wantResp: true, 
        	config: map[string]string{"abckey": "abc val", "key": "val", "key123": "val123"},
        }}
```

File: pkg/qpoption/qpoption_test.go
Comment (psschwei) [inline pkg/qpoption/qpoption_test.go:216]:
```suggestion
	}{{
		name: "empty", 
		wantResp: true, 
		activated: false, 
		config: map[string]string{},
	},{
		name: "other", 
		wantResp: true, 
		activated: false, 
		config: map[string]string{"abckey": "enable"},
	},{
		name: "true", 
		wantResp: true, 
		activated: true, 
		config: map[string]string{myPlugName: "enable"},
	},{
		name: "false", 
		wantResp: true, 
		activated: false, 
		config: map[string]string{myPlugName: "bla"},
	}}
```

File: pkg/qpoption/qpoption_test.go
Comment (psschwei) [inline pkg/qpoption/qpoption_test.go:314]: should we remove this comment? if so

```suggestion
type myDefaultTransport struct {}
```

**Review by psschwei — COMMENTED** [review]
a couple of nits on formatting, etc. but otherwise LGTM
/assign @maximilien

File: README.md
Comment (maximilien) [inline README.md:5]: I would use a bullet list for the different types of vulnerabilities. Just so that its easy to see and improve (like adding more)

File: README.md
Comment (maximilien) [inline README.md:11]: This (one below) is where a diagram summarizing flow could help?

Comment (maximilien) [reply README.md:11]: Actually if you do a diagram, move it right before the next paragraph since that diagram likely would include the different components you are describing below.

File: README.md
Comment (maximilien) [inline README.md:33]: Any chance for a sequence diagram that shows how the vulnerability is trapped and handled. Essentially the components interactions...

File: README.md
Comment (maximilien) [inline README.md:77]: Is there a type in the IBM Research link?

File: pkg/pluginterfaces/pluginterfaces.go
Comment (maximilien) [inline pkg/pluginterfaces/pluginterfaces.go:28]: Do we need to define our own interface for logging? I am guessing (could be wrong) that an existing logging interface exist in serving or Knative pkg?

**Review by maximilien — COMMENTED** [review]
Any way to have shorter PRs? This one seems like it could be broken into smaller. Definitely the dependencies could be their own PR?

I ask because this creates a burden for reviewer. Or at a minimum means multiple day reviews...

File: pkg/pluginterfaces/pluginterfaces.go
Comment (davidhadas) [reply pkg/pluginterfaces/pluginterfaces.go:28]: True, Knative uses zap logger. 
But the security plugs are designed to enable using them in more generic Kubernetes use cases where other loggers may be used. The approach taken is to define a logger interface (that is much needed and unfortunately missing in golang)

File: README.md
Comment (davidhadas) [reply README.md:11]: Please see [my fork](https://github.com/davidhadas/security-guard/tree/addQPOptionAndTestGate) - there is a diagram.

File: README.md
Comment (davidhadas) [reply README.md:33]: There is a diagram that shows the components and their interactions.

File: README.md
Comment (davidhadas) [reply README.md:77]: The guardians.wsecurity.ibmresearch.com is the CRD that is used by Guard:
```
kubectl get crds
NAME                                                  CREATED AT
...
certificates.networking.internal.knative.dev          2022-07-13T17:15:01Z
...
clusterdomainclaims.networking.internal.knative.dev   2022-07-13T17:15:03Z
...
configurations.serving.knative.dev                    2022-07-13T17:15:02Z
domainmappings.serving.knative.dev                    2022-07-13T17:15:03Z
...
guardians.wsecurity.ibmresearch.com                   2022-07-30T06:54:39Z
...
revisions.serving.knative.dev                         2022-07-13T17:15:05Z
routes.serving.knative.dev                            2022-07-13T17:15:06Z
serverlessservices.networking.internal.knative.dev    2022-07-13T17:15:06Z
services.serving.knative.dev                          2022-07-13T17:15:07Z
...
```

And than there is a CRD object to any monitored service (that is if CRDs are used and not Configmaps)
```
kubectl get guardians.wsecurity.ibmresearch.com
NAME            AGE
helloworld-go   10d
mytestservice   9d
```

**Comment by davidhadas** [comment]
Split: 
README files review was moved to https://github.com/knative-sandbox/security-guard/pull/6
Structure and dependencies review was moved to https://github.com/knative-sandbox/security-guard/pull/5
This PR is on hold until PR/5 is merged

File: pkg/pluginterfaces/pluginterfaces.go
Comment (davidhadas) [reply pkg/pluginterfaces/pluginterfaces.go:57]: Actually, this is used today by the reader - i.e. qpoption which supports a single plug and reads RoundTripPlugs[0] (or rtplugs which support multiple plugs and loop through the list of available plugs). 

We can wrap it with GetRegisteredPlugs()  []RoundTripPlug which will return a copy of the RoundTripPlugs and then make RoundTripPlugs not exposed.

File: pkg/qpoption/qpoption.go
Comment (davidhadas) [reply pkg/qpoption/qpoption.go:140]: The rtplugs package supports multiple plugs
I avoided it since it was a bit too much for this repo. 

Originally, the plan was for two repos - one for generic extendibility of QP with qpoption, rtplugs, testgate and pluginterfaces , and another for security-related codes

Since we start with a single repo, I am attempting to remove the support for multiple plugs to somewhat reduce the code complexity - while assuming that we will move back to rtplugs later on.  

>> I will add a comment

File: pkg/pluginterfaces/pluginterfaces.go
Comment (davidhadas) [reply pkg/pluginterfaces/pluginterfaces.go]: Added GetPlug() and GetPlugByName() and moved to non exported list

File: pkg/qpoption/qpoption.go
Comment (davidhadas) [reply pkg/qpoption/qpoption.go]: Solved in pluginterfaces  - see GetPlug()

**Review by maximilien — APPROVED** [review]
/approve
/lgtm

**Pull request merged**
